### PR TITLE
Adding Torch NOTMAD (Dev PR)

### PR DESCRIPTION
**Pull request by wtt102**
- Now uses Explainer & NGAM from modules.py
- Started using masking local to torch notmad class

### CODE REVIEW

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+            "lr_scheduler": {
+                "scheduler": sch,
+                "monitor": "train_loss",
+            },
+        }
+
+    def training_step(self, batch, batch_idx):
+        C, x_true = batch
+        w_pred = self.forward(C).float()
+        loss = self.my_loss(x_true.float(), self._mask(w_pred)).float()
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:123]: It looks like `_mask` is only called during training, so this will output whatever the archetype diagonals were initialized with before training. We might want to set the archetype diagonals to 0 before training as well, you could do `explainer.set_archetypes(self._mask(explainer.get_archetypes()))` in `__init__`.

**Review by cnellington — CHANGES_REQUESTED** [review]
Really really excellent work, I'm thoroughly impressed. In addition to the small changes I requested, please merge in the recent updates from our dev branch so we can see the new CI tests passing on this pull request. I think we'll be able to merge after these changes.

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    # helpers
+    def _mask(self, W):
+        d = W.shape[1]
+        mask = torch.ones(d).float() - torch.eye(d).float()
+        return torch.multiply(W, mask)
+
+    def _parse_alpha_rho(self, params):
+        alpha = params["alpha"]
+        rho = params["rho"]
+        dynamic = False
```
Comment (blengerich) [inline contextualized/dags/torch_notmad/torch_notmad.py:170]: Should this be a parameter?

Comment (cnellington) [reply contextualized/dags/torch_notmad/torch_notmad.py:170]: It looks like DynamicAlphaRho has to be specified outside NOTAMD in the Trainer right now. We should probably make this internal to NOTMAD if possible instead of spreading it over two objects.

Comment (wtt102) [reply contextualized/dags/torch_notmad/torch_notmad.py]: Looked into this and was actually just able to get dynamic alpha rho to now work as not a callback (in most recent commit).

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                )
+            except:
+                print("Error, couldn't project to dag. Returning normal predictions.")
+
+        return w_preds
+
+    # helpers
+    def _mask(self, W):
+        d = W.shape[1]
+        mask = torch.ones(d).float() - torch.eye(d).float()
```
[the remark this answers is not in the document]
Comment (wtt102) [reply contextualized/dags/torch_notmad/torch_notmad.py]: Just setup the mask to be initialized in the init function. Just wanted to clarify if this was the fix here (as seen in your code)?

[the remark this answers is not in the document]
Comment (cnellington) [reply contextualized/dags/torch_notmad/torch_notmad.py]: So sorry, but I linked to some old code that will also cause that error. The correct version would have an `__init__` with
`self.register_buffer("diag_mask", torch.ones(x_dim, x_dim) - torch.eye(x_dim))`
which lets you call `self.diag_mask` on the initialized device without specifying it as a parameter.

[the remark this answers is not in the document]
Comment (cnellington) [reply contextualized/dags/torch_notmad/torch_notmad.py]: See here: https://github.com/cnellington/Contextualized/blob/39cd91ff9888aaaac518a93525f57f48b9503fbe/contextualized/regression/lightning_modules.py#L384

**Review by blengerich — APPROVED** [review]
LGTM!

Some style changes could be made, but I think we are ok to merge this and then make another commit with python black style updates.

**Review by cnellington — APPROVED** [review]
Excellent! Merging now, can't wait to use this!
